addon/mixins/callback-route-mixin.js

```diff
@@ -0,0 +1,67 @@
+import Ember from 'ember';
+
+const { inject: { service }, Mixin, isEmpty } = Ember;
+
+export default Mixin.create({
+  /**
+   The session service.
+
+   @property session
+   @readOnly
+   @type SessionService
+   @public
+   */
+  session: service('session'),
+
+  /**
+   The authenticator that should be used to authenticate the callback.
+
+   @property authenticator
+   @type String
+   @default null
+   @public
+   */
+  authenticator: null,
+
+  /**
+   The error that ocurred during the authentication attempt.
+
+   @property error
+   @type String
+   @default null
+   @public
+   */
+  error: null,
+
+  /**
+   The route that should be transitioned to on error.
+
+   @property routeOnError
+   @type String
+   @default null
+   @public
+   */
+  routeOnError: null,
+
+  /**
+   Authenticates the passed authentication server response and sets the error
+   property if the authentication failed.
+
+   If the routeOnError property is set it will transition to the set route with
+   the error code specified as a query parameter.
+
+   @method activate
+   @public
+   */
+  activate() {
+    let authenticator = this.get('authenticator');
+
+    return this.get('session').authenticate(authenticator).catch((err) => {
+      this.set('error', err);
+
+      if (!isEmpty(this.get('routeOnError'))) {
+        return this.transitionTo(this.get('routeOnError'), { queryParams: { error: err } });
```

Comment:
The point is it's easy to add a transition in an override if you need it but it's harder to remove it if you don't need it.

I think what we want to achieve with the mixin is to encapsulate the code that **everybody** who uses this would have to use but leave everything else out. The problem is that once you start implementing configurable behavior there will be people demanding configuration options for other things as well and if you actually implement all these configuration options you end up with a huge and complex mess. The general strategy of ESA always has been to not add a lot of configurable behavior but make it easy to add it if you need it by overriding methods.